src/groth16/prover.rs

```diff
@@ -310,152 +331,167 @@ where
         None
     };
 
-    let mut fft_kern = Some(LockedFFTKernel::<E>::new(log_d, priority));
+    let mut a_s = Vec::with_capacity(num_circuits);
+    let mut params_h = None;
+    let worker = &worker;
+    let provers_ref = &mut provers;
+    let params = &params;
 
-    let a_s = provers
-        .iter_mut()
-        .map(|prover| {
-            let mut a =
-                EvaluationDomain::from_coeffs(std::mem::replace(&mut prover.a, Vec::new()))?;
-            let mut b =
-                EvaluationDomain::from_coeffs(std::mem::replace(&mut prover.b, Vec::new()))?;
-            let mut c =
-                EvaluationDomain::from_coeffs(std::mem::replace(&mut prover.c, Vec::new()))?;
-
-            a.ifft(&worker, &mut fft_kern)?;
-            a.coset_fft(&worker, &mut fft_kern)?;
-            b.ifft(&worker, &mut fft_kern)?;
-            b.coset_fft(&worker, &mut fft_kern)?;
-            c.ifft(&worker, &mut fft_kern)?;
-            c.coset_fft(&worker, &mut fft_kern)?;
-
-            a.mul_assign(&worker, &b);
-            drop(b);
-            a.sub_assign(&worker, &c);
-            drop(c);
-            a.divide_by_z_on_coset(&worker);
-            a.icoset_fft(&worker, &mut fft_kern)?;
-            let mut a = a.into_coeffs();
-            let a_len = a.len() - 1;
-            a.truncate(a_len);
-
-            Ok(Arc::new(
-                a.into_iter().map(|s| s.0.to_repr()).collect::<Vec<_>>(),
-            ))
-        })
-        .collect::<Result<Vec<_>, SynthesisError>>()?;
+    THREAD_POOL.scoped(|s| {
+        let a_s = &mut a_s;
+        s.execute(move || {
+            let mut fft_kern = Some(LockedFFTKernel::<E>::new(log_d, priority));
 
-    drop(fft_kern);
-    let mut multiexp_kern = Some(LockedMultiexpKernel::<E>::new(log_d, priority));
+            for prover in provers_ref {
+                a_s.push(execute_fft(worker, prover, &mut fft_kern));
+            }
+        });
+
+        debug!("get h");
+        params_h = Some(params.get_h(n));
```

Comment:
In this case you do the kernel creation and the parameter loading in parallel. You spawn the kernel creation into a new thread, but you wait until both are finished. In the code below, you spawn the parameter generation and do something else on the main thread. Would it make sense (for consistency, and easier to follow code) to change this part to also spawn the parameter generation and leave the kernel generation on the main thread?

Reply:
yes, fixed